### DISCUSSION TITLE
Transient state nodes run in O(n) time 😞

### DISCUSSION THREAD

**Discussion by awreccan** [statelyai/xstate#2096] [Q&A] [answered]
I have some performance concerns about [transient state nodes](https://xstate.js.org/docs/guides/statenodes.html#transient-state-nodes). Would love some clarification.

### Background
I created a machine with a history stack in it: https://github.com/davidkpiano/xstate/issues/188#issuecomment-819095230. This enables having a "back button" on every state, allowing you to retrace the path you took in the state machine from the initial node.

The implementation relies on a transient state node called "back", which maintains a "registry" of all the nodes it could _eventlessly_ transition the machine to.
```js
      back: {
        always: Object.keys(allStates).map((state) => {
          return {
            target: state,
            actions: updateHistoryOnBack,
            cond: isPrevState(state)
          };
        })
      }
```
visually, you can see the "registry" maintained by the transient state node, "back":
![image](https://user-images.githubusercontent.com/10543509/114817732-221bf680-9ded-11eb-928c-c76bcca79cd2.png)

### Problem: O(n) vs O(1)
> ![image](https://user-images.githubusercontent.com/10543509/114817168-1845c380-9dec-11eb-960f-a14459d9761f.png)
> from https://xstate.js.org/docs/guides/transitions.html#eventless-always-transitions

My concern is that the back node runs _all_ `always` `cond`s to find the correct one. This seems unnecessarily computationally expensive to me (O(n)). All I really want to do is to read `prevState = history.peek()` and navigate to it.

An O(1) solution would be possible if `always` supported syntax like:
```js
      back: {
        always: [{
          pick: (ctx) => getPrevState(ctx), // this "pick" key tells xstate which key to pick from this object (keys represent transition targets)
          actions: { ... }, // actions on transition from this transient state node to another node

          // possible states this transient state node could transition to
          state1: { actions: updateHistoryOnBack }, // actions specific to...
          state2: { actions: updateHistoryOnBack }, // ...a particular picked target
          ...
        }]
      }
```

### Discussion
Since I'm looking at a statechart that could be quite large, I'd love to understand:
- Is it possible to do this in O(1) time with the current syntax?
- Is it possible to introduce O(1) syntax like that suggested above?

**Comment by Andarist** [comment, marked as answer]
> Is it possible to do this in O(1) time with the current syntax?

No, not right now.

> Is it possible to introduce O(1) syntax like that suggested above?

This would be a somewhat heavy addition for a rare use case so I would say that: no, currently there is no plan for something like this. Keep in mind that it's not as simple as adding some additional syntax - we'd have to think through how to visualize this, how this fits the prior art, and if it's even a use case that we want to support. We might reevaluate this in the future though.

---

Instead of maintaining a single transient state, I would refactor this to keep a single transient state per your "main" state. You can transform those main states into compound states which would make it easier to implement and handle.

Also - how often `BACK` is send to your machine? Even if the statechart gets pretty big - how much time it will take the runner to check all conditions? For this to become a bottleneck your statechart would have to be enormous. Please don't consider only theoretical Big O complexity but also actual execution times.

**Reply by awreccan** [reply]
Thanks for the response!

> Instead of maintaining a single transient state, I would refactor this to keep a single transient state per your "main" state. You can transform those main states into compound states which would make it easier to implement and handle.

I don't follow, could you explain in greater detail?

> Please don't consider only theoretical Big O complexity but also actual execution times.

Fair enough. At the moment, the statechart is manageable, but I do envision future use cases where we plug multiple flows 
together (by leveraging the ability to nest statecharts infinitely). But we'll get there when we get there.

**Reply by Andarist** [reply]
>  I do envision future use cases where we plug multiple flows together (by leveraging the ability to nest statecharts infinitely). But we'll get there when we get there.

Sure, this is possible but I would probably encourage you to split things up a little bit - with a single huge statechart you might end up with event name collisions easier, the visualization will be huge etc. But it makes it easier to share events between regions - so I see why in certain situations you could want this.

> I don't follow, could you explain in greater detail?

Each state could look smth like this:
```js
{
    foo: {
        initial: 'regular',
        states: {
            regular: {
                on: {
                    BACK: 'back'
                }
            },
            back: {
                always: [/* generate back transitions */]
            }
        }
    }
}
```

**Reply by awreccan** [reply]
Thank you for your ideas!

> I would probably encourage you to split things up a little bit

Noted!

> Each state could look smth like this

Having a transient node per state would entail replicating the always transitions (numbering as many as nodes in the graph) per state - n^2 transitions. You might ask why every state needs the ability to transition back to every other state. A little context would be helpful in answering that - I'm creating a generalised back button library, which can then give developers the freedom to write their own statecharts with special flags on states to indicate that a back button should be present on that state. So at a library level, I must allow transitions back to any of the n states in the statechart without presuming that you can only go back to a subset of states. Of course I could improve this by some reachability analysis on the graph (can't go back from Y to X if you can't reach Y from X), but that'll still result in >n always transitions. Is there any other reason why a per-node transient back state sounds better to you, given this context?
